File: rust-client/src/math/curve.rs
use cp_amm_interface::Rounding;
use ruint::aliases::U256;

use anyhow::{ensure, Context, Ok, Result};

use crate::math::mul_div_u256;

pub const RESOLUTION: u8 = 64;

pub fn get_initialize_amounts(
    sqrt_min_price: u128,
    sqrt_max_price: u128,
    sqrt_price: u128,
    liquidity: u128,
) -> Result<(u64, u64)> {
    // BASE TOKEN
    let amount_a =
        get_delta_amount_a_unsigned(sqrt_price, sqrt_max_price, liquidity, Rounding::Up)?;
    // QUOTE TOKEN
    let amount_b =
        get_delta_amount_b_unsigned(sqrt_min_price, sqrt_price, liquidity, Rounding::Up)?;
    Ok((amount_a, amount_b))
}

/// Gets the delta amount_a for given liquidity and price range
///
/// # Formula
///
/// * `Δa = L * (1 / √P_lower - 1 / √P_upper)`
/// * i.e. `L * (√P_upper - √P_lower) / (√P_upper * √P_lower)`
pub fn get_delta_amount_a_unsigned(
    lower_sqrt_price: u128,
    upper_sqrt_price: u128,
    liquidity: u128,
    round: Rounding,
) -> Result<u64> {
    let result = get_delta_amount_a_unsigned_unchecked(
        lower_sqrt_price,
        upper_sqrt_price,
        liquidity,
        round,
    )?;
    ensure!(result <= U256::from(u64::MAX), "MathMathOverflow");

    return Ok(result.try_into().context("CastTypeFailed")?);
}

/// * i.e. `L * (√P_upper - √P_lower) / (√P_upper * √P_lower)`
pub fn get_delta_amount_a_unsigned_unchecked(
    lower_sqrt_price: u128,
    upper_sqrt_price: u128,
    liquidity: u128,
    round: Rounding,
) -> Result<U256> {
    let numerator_1 = U256::from(liquidity);
    let numerator_2 = U256::from(upper_sqrt_price - lower_sqrt_price);

    let denominator = U256::from(lower_sqrt_price)
        .checked_mul(U256::from(upper_sqrt_price))
        .context("MathOverflow")?;

    assert!(denominator > U256::ZERO);
    let result =
        mul_div_u256(numerator_1, numerator_2, denominator, round).context("MathMathOverflow")?;
    return Ok(result);
}

/// Gets the delta amount_b for given liquidity and price range
/// * `Δb = L (√P_upper - √P_lower)`
pub fn get_delta_amount_b_unsigned(
    lower_sqrt_price: u128,
    upper_sqrt_price: u128,
    liquidity: u128,
    round: Rounding,
) -> Result<u64> {
    let result = get_delta_amount_b_unsigned_unchecked(
        lower_sqrt_price,
        upper_sqrt_price,
        liquidity,
        round,
    )?;
    ensure!(result <= U256::from(u64::MAX), "MathMathOverflow");
    return Ok(result.try_into().context("TypeCastFailed")?);
}

//Δb = L (√P_upper - √P_lower)
pub fn get_delta_amount_b_unsigned_unchecked(
    lower_sqrt_price: u128,
    upper_sqrt_price: u128,
    liquidity: u128,
    round: Rounding,
) -> Result<U256> {
    let liquidity = U256::from(liquidity);
    let delta_sqrt_price = U256::from(upper_sqrt_price - lower_sqrt_price);
    let prod = liquidity
        .checked_mul(delta_sqrt_price)
        .context("MathMathOverflow")?;

    match round {
        Rounding::Up => {
            let denominator = U256::from(1)
                .checked_shl((RESOLUTION as usize) * 2)
                .context("MathOverflow")?;
            let result = prod.div_ceil(denominator);
            Ok(result)
        }
        Rounding::Down => {
            let (result, _) = prod.overflowing_shr((RESOLUTION as usize) * 2);
            Ok(result)
        }
    }
}

/// Gets the next sqrt price given an input amount of token_a or token_b
/// Throws if price or liquidity are 0, or if the next price is out of bounds
pub fn get_next_sqrt_price_from_input(
    sqrt_price: u128,
    liquidity: u128,
    amount_in: u64,
    a_for_b: bool,
) -> Result<u128> {
    assert!(sqrt_price > 0);
    assert!(liquidity > 0);

    // round to make sure that we don't pass the target price
    if a_for_b {
        get_next_sqrt_price_from_amount_a_rounding_up(sqrt_price, liquidity, amount_in)
    } else {
        get_next_sqrt_price_from_amount_b_rounding_down(sqrt_price, liquidity, amount_in)
    }
}

pub fn get_next_sqrt_price_from_amount_a_rounding_up(
    sqrt_price: u128,
    liquidity: u128,
    amount: u64,
) -> Result<u128> {
    if amount == 0 {
        return Ok(sqrt_price);
    }
    let sqrt_price = U256::from(sqrt_price);
    let liquidity = U256::from(liquidity);

    let product = U256::from(amount)
        .checked_mul(sqrt_price)
        .context("MathOverflow")?;
    let denominator = liquidity
        .checked_add(U256::from(product))
        .context("MathOverflow")?;
    let result =
        mul_div_u256(liquidity, sqrt_price, denominator, Rounding::Up).context("MathOverflow")?;
    return Ok(result.try_into().context("TypeCastFailed")?);
}

pub fn get_next_sqrt_price_from_amount_b_rounding_down(
    sqrt_price: u128,
    liquidity: u128,
    amount: u64,
) -> Result<u128> {
    let quotient = U256::from(amount)
        .checked_shl((RESOLUTION * 2) as usize)
        .context("MathOverflow")?
        .checked_div(U256::from(liquidity))
        .context("MathOverflow")?;

    let result = U256::from(sqrt_price)
        .checked_add(quotient)
        .context("MathOverflow")?;
    Ok(result.try_into().context("TypeCastFailed")?)
}

File: src/math/curve.ts
import BN from "bn.js";
import { mulDiv } from "./utilsMath";
import { SCALE_OFFSET, U64_MAX } from "../constants";
import { Rounding } from "../types";

/**
 * Gets the next sqrt price given a delta of token_b (input), rounding down.
 * √P' = √P + Δb / L
 * @param sqrtPrice - current sqrt price (BN)
 * @param liquidity - current liquidity (BN)
 * @param amount - delta of token_b (BN)
 * @returns next sqrt price (BN)
 */
export function getNextSqrtPriceFromAmountInBRoundingDown(
  sqrtPrice: BN,
  liquidity: BN,
  amount: BN
): BN {
  // quotient = (amount << (SCALE_OFFSET * 2)) / liquidity
  const quotient = amount.shln(SCALE_OFFSET * 2).div(liquidity);

  // result = sqrtPrice + quotient
  const result = sqrtPrice.add(quotient);

  return result;
}

/**
 * Gets the next sqrt price given a delta of token_b (output), rounding down.
 * √P' = √P - Δb / L
 * @param sqrtPrice - current sqrt price (BN)
 * @param liquidity - current liquidity (BN)
 * @param amount - delta of token_b (BN)
 * @returns next sqrt price (BN)
 */
export function getNextSqrtPriceFromAmountOutBRoundingDown(
  sqrtPrice: BN,
  liquidity: BN,
  amount: BN
): BN {
  // quotient = ceil((amount << (SCALE_OFFSET * 2)) / liquidity)
  const numerator = amount.shln(SCALE_OFFSET * 2);
  const quotient = numerator.add(liquidity).subn(1).div(liquidity);

  // result = sqrtPrice - quotient
  const result = sqrtPrice.sub(quotient);

  if (result.isNeg()) {
    throw new Error("sqrt price cannot be negative");
  }

  return result;
}

/**
 * Gets the next sqrt price √P' given a delta of token_a, rounding up.
 * √P' = √P * L / (L + Δa * √P)
 * @param sqrtPrice - current sqrt price (BN)
 * @param liquidity - current liquidity (BN)
 * @param amount - delta of token_a (BN)
 * @returns next sqrt price (BN)
 */
export function getNextSqrtPriceFromAmountInARoundingUp(
  sqrtPrice: BN,
  liquidity: BN,
  amount: BN
): BN {
  if (amount.isZero()) {
    return sqrtPrice;
  }

  // product = amountIn * sqrtPrice
  const product = amount.mul(sqrtPrice);

  // denominator = liquidity + product
  const denominator = liquidity.add(product);

  // result = mulDiv(liquidity, sqrtPrice, denominator, Rounding.Up)
  const result = mulDiv(liquidity, sqrtPrice, denominator, Rounding.Up);

  return result;
}

/**
 * Gets the next sqrt price √P' given a delta of token_a (output), rounding up.
 * √P' = √P * L / (L - Δa * √P)
 * @param sqrtPrice - current sqrt price (BN)
 * @param liquidity - current liquidity (BN)
 * @param amount - delta of token_a (BN)
 * @returns next sqrt price (BN)
 */
export function getNextSqrtPriceFromAmountOutARoundingUp(
  sqrtPrice: BN,
  liquidity: BN,
  amount: BN
): BN {
  if (amount.isZero()) {
    return sqrtPrice;
  }

  // product = amount * sqrtPrice
  const product = amount.mul(sqrtPrice);

  const denominator = liquidity.sub(product);
  if (denominator.lte(new BN(0))) {
    throw new Error("MathOverflow: denominator is zero or negative");
  }

  // result = mulDiv(liquidity, sqrtPrice, denominator, Rounding.Up)
  return mulDiv(liquidity, sqrtPrice, denominator, Rounding.Up);
}

/**
 * Gets the next sqrt price given an output amount of token_a or token_b.
 * @param sqrtPrice - current sqrt price (BN)
 * @param liquidity - current liquidity (BN)
 * @param amountOut - output amount (BN)
 * @param aForB - true if swapping A for B, false if B for A
 * @returns next sqrt price (BN)
 */
export function getNextSqrtPriceFromOutput(
  sqrtPrice: BN,
  liquidity: BN,
  amountOut: BN,
  aForB: boolean
): BN {
  if (sqrtPrice.lte(new BN(0))) {
    throw new Error("sqrtPrice must be greater than 0");
  }
  if (liquidity.lte(new BN(0))) {
    throw new Error("liquidity must be greater than 0");
  }

  if (aForB) {
    return getNextSqrtPriceFromAmountOutBRoundingDown(
      sqrtPrice,
      liquidity,
      amountOut
    );
  } else {
    return getNextSqrtPriceFromAmountOutARoundingUp(
      sqrtPrice,
      liquidity,
      amountOut
    );
  }
}

/**
 * Gets the next sqrt price given an input amount of token_a or token_b
 * @param sqrtPrice - current sqrt price (BN)
 * @param liquidity - current liquidity (BN)
 * @param amountIn - input amount (BN)
 * @param aForB - true if swapping A for B, false if B for A
 * @returns next sqrt price (BN)
 */
export function getNextSqrtPriceFromInput(
  sqrtPrice: BN,
  liquidity: BN,
  amountIn: BN,
  aForB: boolean
): BN {
  if (sqrtPrice.lte(new BN(0))) {
    throw new Error("sqrtPrice must be greater than 0");
  }
  if (liquidity.lte(new BN(0))) {
    throw new Error("liquidity must be greater than 0");
  }

  if (aForB) {
    // Rounding up for A to B
    return getNextSqrtPriceFromAmountInARoundingUp(
      sqrtPrice,
      liquidity,
      amountIn
    );
  } else {
    // Rounding down for B to A
    return getNextSqrtPriceFromAmountInBRoundingDown(
      sqrtPrice,
      liquidity,
      amountIn
    );
  }
}

/**
 * Gets the delta amount of token_b for given liquidity and price range. (getDeltaAmountBUnsigned)
 * @param lowerSqrtPrice - lower sqrt price (BN)
 * @param upperSqrtPrice - upper sqrt price (BN)
 * @param liquidity - current liquidity (BN)
 * @param rounding - rounding (Rounding)
 * @returns delta amount of token_b (BN)
 */
export function getAmountBFromLiquidityDelta(
  lowerSqrtPrice: BN,
  upperSqrtPrice: BN,
  liquidity: BN,
  rounding: Rounding
): BN {
  const result = getDeltaAmountBUnsignedUnchecked(
    lowerSqrtPrice,
    upperSqrtPrice,
    liquidity,
    rounding
  );

  if (result.gt(U64_MAX)) {
    throw new Error("MathOverflow: result exceeds u64 max");
  }

  return result;
}

/**
 * Gets the delta amount of token_b for given liquidity and price range.
 * @param lowerSqrtPrice - lower sqrt price (BN)
 * @param upperSqrtPrice - upper sqrt price (BN)
 * @param liquidity - current liquidity (BN)
 * @param rounding - rounding (Rounding)
 * @returns delta amount of token_b (BN)
 */
function getDeltaAmountBUnsignedUnchecked(
  lowerSqrtPrice: BN,
  upperSqrtPrice: BN,
  liquidity: BN,
  rounding: Rounding
): BN {
  // delta_sqrt_price = upper_sqrt_price - lower_sqrt_price
  const deltaSqrtPrice = upperSqrtPrice.sub(lowerSqrtPrice);
  const prod = liquidity.mul(deltaSqrtPrice);

  const shift = SCALE_OFFSET * 2;

  if (rounding === Rounding.Up) {
    // denominator = 1 << (SCALE_OFFSET * 2)
    const denominator = new BN(1).ushln(shift);
    // result = ceil(prod / denominator)
    const result = prod.add(denominator.subn(1)).div(denominator);
    return result;
  } else {
    // result = prod >> (RESOLUTION * 2)
    const result = prod.ushrn(shift);
    return result;
  }
}

/**
 * Gets the delta amount_a for given liquidity and price range. (getDeltaAmountAUnsigned)
 * `Δa = L * (1 / √P_lower - 1 / √P_upper)`
 * i.e. `L * (√P_upper - √P_lower) / (√P_upper * √P_lower)`
 * @param lowerSqrtPrice - lower sqrt price (BN)
 * @param upperSqrtPrice - upper sqrt price (BN)
 * @param liquidity - current liquidity (BN)
 * @param rounding - rounding (Rounding)
 * @returns delta amount of token_a (BN)
 */
export function getAmountAFromLiquidityDelta(
  lowerSqrtPrice: BN,
  upperSqrtPrice: BN,
  liquidity: BN,
  rounding: Rounding
): BN {
  const result = getDeltaAmountAUnsignedUnchecked(
    lowerSqrtPrice,
    upperSqrtPrice,
    liquidity,
    rounding
  );

  if (result.gt(U64_MAX)) {
    throw new Error("MathOverflow: result exceeds u64 max");
  }

  return result;
}

/**
 * Computes L * (√P_upper - √P_lower) / (√P_upper * √P_lower)
 * @param lowerSqrtPrice - lower sqrt price (BN)
 * @param upperSqrtPrice - upper sqrt price (BN)
 * @param liquidity - current liquidity (BN)
 * @param rounding - rounding (Rounding)
 * @returns delta amount of token_a (BN)
 */
function getDeltaAmountAUnsignedUnchecked(
  lowerSqrtPrice: BN,
  upperSqrtPrice: BN,
  liquidity: BN,
  rounding: Rounding
): BN {
  const numerator1 = liquidity;
  const numerator2 = upperSqrtPrice.sub(lowerSqrtPrice);

  const denominator = lowerSqrtPrice.mul(upperSqrtPrice);

  if (denominator.lte(new BN(0))) {
    throw new Error("Denominator must be greater than zero");
  }

  const result = mulDiv(numerator1, numerator2, denominator, rounding);
  return result;
}

/**
 * Gets the liquidity delta from amount A
 * Δa = L * (1 / √P_lower - 1 / √P_upper)
 * Δa = L * (√P_upper - √P_lower) / (√P_upper * √P_lower)
 * L = Δa * √P_upper * √P_lower / (√P_upper - √P_lower)
 * @param amountA - The amount of token A
 * @param lowerSqrtPrice - The lower sqrt price
 * @param upperSqrtPrice - The upper sqrt price
 * @returns The liquidity delta from amount A
 */
export function getLiquidityDeltaFromAmountA(
  amountA: BN,
  lowerSqrtPrice: BN, // current sqrt price
  upperSqrtPrice: BN // max sqrt price
): BN {
  const product = amountA.mul(lowerSqrtPrice).mul(upperSqrtPrice); // Q128.128
  const denominator = upperSqrtPrice.sub(lowerSqrtPrice); // Q64.64

  return product.div(denominator);
}

/**
 * Gets the liquidity delta from amount B
 * Δb = L (√P_upper - √P_lower)
 * L = Δb / (√P_upper - √P_lower)
 * @param amountB - The amount of token B
 * @param lowerSqrtPrice - The lower sqrt price
 * @param upperSqrtPrice - The upper sqrt price
 * @returns The liquidity delta from amount B
 */
export function getLiquidityDeltaFromAmountB(
  amountB: BN,
  lowerSqrtPrice: BN, // min sqrt price
  upperSqrtPrice: BN // current sqrt price,
): BN {
  const denominator = upperSqrtPrice.sub(lowerSqrtPrice);
  const product = amountB.shln(128);
  return product.div(denominator);
}
